lib/http/http.c

```diff
@@ -660,7 +664,10 @@ callback_chunkedheader(void * cookie, int status)
 	/* If we found one, handle the line. */
 	if (eolpos != buflen) {
 		/* Parse the chunk length. */
-		clen = strtoull((const char *)buf, NULL, 16);
+		if (PARSENUM_BASE(&clen, (const char *)buf, 16)) {
+			warnp("Invalid chunk length: %s", (const char *)buf);
```

Comment:
Is `buf` a NULL-terminated string?

Reply:
Mao.

Parsing something like "`5\r\n1234`" as `5` worked with the old `strotoull` because the definition is that the conversion happens after splitting the input into three parts, including

> A final string of one or more unrecognized characters, including the terminating NUL character of the input string
> http://pubs.opengroup.org/onlinepubs/9699919799/functions/strtoul.html

Unfortunately, `PARSENUM` specifically checks to make sure that this doesn't happen by passing `eptr`:
https://github.com/Tarsnap/libcperciva/blob/master/util/parsenum.h#L189

As a result, that particular `PARSENUM_BASE()` will always error out with `EINVAL`.  :(

I see three options:
- keep the old `stroull()`, and add an extra check to make sure that `clen` isn't too large.

  I'm not fond of that idea, because the whole logic of `clen_ull =... ; if (clen_ull < SIZE_MAX) clen = (size_t)clen_ull;` is ugly and takes us out of the abstraction level of parsing a chunked header line.

- switch to `PARSENUM_BASE()`, but add another argument to say "don't complain if `eptr != '\0'`".  I think that would be too finicky.

- copy the relevant bytes of `buf` into a temporary string, then `PARSENUM_BASE` that string.

I think the last option is the cleanest, but I suspect that you might prefer the first one.